### DISCUSSION TITLE
import rxfnsb  - my RXC call doesn't work

### DISCUSSION THREAD

**Discussion by Peter-Jacob** [adesutherland/CREXX#376] [Q&A]
I have once again shredded my RXC definition! I ran the CMAKE Release call, which worked fine. When I run now rxc:
```
C:\...\CREXX-Release\compiler\rxc.exe -i C:\...\CREXX\cmake-build-debug\lib\rxfns SHSort
Importing Procedures - Can't read input file
```
My little test REXX
```
/* REXX */
options levelb
import rxfnsb /* This imports the standard library - for length() etc. */
say right("1234",4)
return
```
I know the standard library is not present, but I don't know which module(s) to look for. Is it an rxbin, or rxas file(s). It is definitely not rxfnsb that I am importing. Perhaps the documentation should clarify this, this would make it easier to find the missing library in the environment. 
I also noticed that `import rxfnsb' does not report an error if none of the functions it contains are called. So the existence check is done on the first function call.

**Comment by Peter-Jacob** [comment]
I worked around the problem by cloning f0049 again. The underlying problem, or my misunderstanding, remains. What exactly does the import rxfnsb do, does it look for one big module, does it expect all the levelb functions to be available in the -i parameter?

**Reply by adesutherland** [reply]
Hi Peter. 

The import gets the rxc to search 
.rxbin files. In the project directory (where the rexx file being compiled is), and in the installation directory (where the rxc binary is). You can also add search folders (-l flag but use -h for the help text)

The file it needs to find is library.rxbin which you will find in the build lib folder, I think. 

The easiest is to copy rxc, rxas, rxdas, rxvme, library.rxbin all into some directory with a script and add the directory to PATH. Remembering to rerun the copy script after an update ...

Hope this helps

A

**Reply by Peter-Jacob** [reply]
Thanks Adrian, I already thought library.rxbin might be the module needed. I therefore let the -i parm point to its directory, without success.  Is there any validation of : date/version or debug vs release in the import process?

**Comment by rvjansen** [comment]
Exactly.

But because we will always forget to copy things when needed, and because copying is tedious work in itself, this is what I do:
I build a release build in apps/crexx_release (and a debug build in apps/crexx_debug but that is less relevant here)

The build from f0049 includes the crexx executable from the bin directory. I put this on path:

PATH=~/bin
PATH=$PATH:/Users/rvjansen/apps/crexx_release/assembler
PATH=$PATH:/Users/rvjansen/apps/crexx_release/compiler
PATH=$PATH:/Users/rvjansen/apps/crexx_release/debugger
PATH=$PATH:/Users/rvjansen/apps/crexx_release/disassembler
PATH=$PATH:/Users/rvjansen/apps/crexx_release/interpreter
PATH=$PATH:/Users/rvjansen/apps/crexx_release/cpacker
PATH=$PATH:/Users/rvjansen/apps/crexx_release/bin


➜  ~ crexx
cRexx compiler driver macosx 64 crexx-f0051 20240815
Copyright (c) Adrian Sutherland 2021,2024. All rights reserved.
Copyright (c) RexxLA 2021,2024. All rights reserved.

Arguments are: in_file_specification... [--option]...

The following options are available:

--help           -- display (this) help info
--version        -- display the version number
--exec           -- execute (default)
--native         -- build native executable; implies noexec; default nonative
➜  ~ which crexx
/Users/rvjansen/apps/crexx_release/bin/crexx
➜  ~ echo $CREXX_HOME
/Users/rvjansen/apps/crexx_release
➜  ~

The crexx executable looks into the contents of the CREXX_HOME variable, and then will find all changed things - every time I build them. And it can run the program with rxvm or it can build a native executable using cpacker and the gcc compiler. I imagine we have to work on this for the eventual release, but for me it works all the time.

best regards,

René.

**Reply by rvjansen** [reply]
and while testing that again I noticed that compile to native failed because of a missing rxpa lib; which is now fixed.
For distributions, I think the approach of one directory is best, for development however, I think is it quite dangerous because of the possibility of linking against stale code, which can set you back a few hours.
